flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkV1Adapter.java

```diff
@@ -0,0 +1,447 @@
+        @Override
+        public SinkWriterV1Adapter<InputT, CommT, WriterStateT> createWriter(InitContext context)
+                throws IOException {
+            return SinkV1Adapter.this.createWriter(context);
+        }
+
+        @VisibleForTesting
+        public org.apache.flink.api.connector.sink.Sink<InputT, CommT, WriterStateT, GlobalCommT>
+                getSink() {
+            return sink;
+        }
+    }
+
+    private class StatefulSinkAdapter extends PlainSinkAdapter
+            implements StatefulSink<InputT, WriterStateT> {
+        @Override
+        public StatefulSinkWriter<InputT, WriterStateT> restoreWriter(
+                InitContext context, Collection<WriterStateT> recoveredState) throws IOException {
+            org.apache.flink.api.connector.sink.SinkWriter<InputT, CommT, WriterStateT> writer =
+                    sink.createWriter(
+                            new InitContextAdapter(context), new ArrayList<>(recoveredState));
+            return new SinkWriterV1Adapter<>(writer);
+        }
+
+        @Override
+        public SimpleVersionedSerializer<WriterStateT> getWriterStateSerializer() {
+            return sink.getWriterStateSerializer()
+                    .orElseThrow(
+                            () ->
+                                    new IllegalStateException(
+                                            "This method should only be called after adapter established that the result is non-empty."));
+        }
+    }
+
+    private class TwoPhaseCommittingSinkAdapter extends PlainSinkAdapter
+            implements TwoPhaseCommittingSink<InputT, CommT>, WithCompatibleState {
+        @Override
+        public Committer<CommT> createCommitter() throws IOException {
+            return new CommitterAdapter<>(
+                    sink.createCommitter().orElse(new SinkV1Adapter.NoopCommitter<>()));
+        }
+
+        @Override
+        public SimpleVersionedSerializer<CommT> getCommittableSerializer() {
+            return sink.getCommittableSerializer()
+                    .orElseThrow(
+                            () ->
+                                    new IllegalStateException(
+                                            "This method should only be called after adapter established that the result is non-empty."));
+        }
+
+        @Override
+        public Collection<String> getCompatibleWriterStateNames() {
+            return sink.getCompatibleStateNames();
+        }
+    }
+
+    private class GlobalCommittingSinkAdapter extends TwoPhaseCommittingSinkAdapter
+            implements WithPostCommitTopology<InputT, CommT> {
+
+        @Override
+        public void addPostCommitTopology(DataStream<CommittableMessage<CommT>> committables) {
+            StandardSinkTopologies.addGlobalCommitter(committables, GlobalCommitterAdapter::new);
+        }
+    }
+
+    private class StatefulTwoPhaseCommittingSinkAdapter extends StatefulSinkAdapter
```

Comment:
I am not sure I understand your idea. The problem is that previously we only had one interface `Sink<InputT, CommT, WriterStateT, GlobalCommT>`  that is now represented by multiple mixins. So I had to split the different use cases into different classes implementing the new interfaces.

Reply:
ISP(Interface Segregation Principle) is controlled by the interface not the implementation classes. It is fine to split the big old Sink interface based on ISP. Using one class to implement them all will not violate it. Because what the caller will see are fine-grained interfaces not the implementation class(es), especially when all of these classes are private inner classes and most logic is to wrap each other.

Reply:
AFAICT I need multiple classes or it increases the complexity to implement the given use cases with one class i.e. converting a previously only SinkWriter -> GlobalCommitter topologies now becomes SinkWriter -> ForwardCommitter -> GlobalCommitter. I think the conditional logic is then distributed across all creation methods (i.e. SinkWriter, Commiter, GlobalCommiter) which I find harder to understand instead of doing the checks once and returning different implementations.